### PR TITLE
[RFR] v8 api, guidelines

### PR DESCRIPTION
**Pull request by hbrunn**
This is a refactoring for `partner_relations` because in https://github.com/OCA/partner-contact/pull/264#issuecomment-239178339, we noticed it doesn't make very much sense to migrate this module to 9.0 when there's still a couple of things wrong with the 8.0 version.

### CODE REVIEW

File: partner_relations/models/res_partner.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
+# -*- coding: utf-8 -*-
+# © 2013-2016 Therp BV <http://therp.nl>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+import numbers
+from openerp import _, models, fields, exceptions, api
```
Comment (NL66278) [inline partner_relations/models/res_partner.py]: Alphabetic ordering?

**Comment by NL66278** [comment]
:-1: Had to change this. Code looks good, but still strange things are happening when adding relation.

Way to reproduce (in db with demo data):
- Go into a company
- Use button to go to relations
- Add has former employee relation and relate to a person
- After saving the relation shown is Has worked for and the Other partner is actually the organisation originally selected

This is a display problem. After going back to the relation (using the breadcrumb for instance) and then again using the button to go to relations, the data is shown as should be expected.

File: partner_relations/models/res_partner.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+    )
+
+    @api.one
+    @api.depends("relation_ids")
+    def _compute_relation_count(self):
+        """Count the number of relations this partner has for Smart Button
+
+        Don't count inactive relations.
+        """
+        self.relation_count = len([r for r in self.relation_ids if r.active])
```
Comment (pedrobaeza) [inline partner_relations/models/res_partner.py]: Use `len(self.relation_ids.filtered('active'))`, but I think this is not even needed, as ORM by default hides non active records, so `len(self.relation_ids)` should be enough.

File: partner_relations/models/res_partner.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+
+    @api.multi
+    def _compute_relation_ids(self):
+        '''getter for relation_ids'''
```
Comment (pedrobaeza) [inline partner_relations/models/res_partner.py]: Use double quote

File: partner_relations/models/res_partner.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+        )
+        partner2relation = dict(self.env.cr.fetchall())
+        for this in self:
+            this.relation_ids += self.env['res.partner.relation'].browse(
+                partner2relation.get(this.id, []),
+            )
+
+    @api.model
+    def _search_relation_id(self, operator, value):
+        result = []
```
Comment (pedrobaeza) [inline partner_relations/models/res_partner.py:67]: Can you please remove and contract a bit these methods? They have a lot of empty lines and the brackets are very expanded so the readability is hard.

File: partner_relations/models/res_partner_relation.py [hunk 5]
```diff
@@ -41,73 +22,40 @@ class ResPartnerRelation(models.Model):
...
-    def _search_any_partner_id(self, operator, value):
-        return [
-            '|',
-            ('left_partner_id', operator, value),
-            ('right_partner_id', operator, value),
-        ]
+    type_selection_id = fields.Many2one(
+        'res.partner.relation.type.selection',
+        compute='_compute_fields',
+        fnct_inv=lambda *args: None,
```
Comment (pedrobaeza) [inline partner_relations/models/res_partner_relation.py]: You don't need to indicate inverse function

File: partner_relations/models/res_partner_relation.py [hunk 6]
```diff
@@ -138,55 +86,111 @@ def get_values(self, dummy_field_names, dummy_arg, context=None):
...
+                        (on_right_partner and 1 or 0))
+            this.partner_id_display = (
+                this.left_partner_id
+                if on_right_partner
+                else this.right_partner_id
+            )
+
+    @api.onchange('type_selection_id')
+    def _onchange_type_selection_id(self):
+        '''Set domain on partner_id_display, when selection a relation type'''
```
Comment (pedrobaeza) [inline partner_relations/models/res_partner_relation.py]: Double quote

File: partner_relations/models/res_partner_relation.py [hunk 7]
```diff
@@ -138,55 +86,111 @@ def get_values(self, dummy_field_names, dummy_arg, context=None):
...
+    def _search_any_partner_id(self, operator, value):
+        return [
+            '|',
+            ('left_partner_id', operator, value),
+            ('right_partner_id', operator, value),
+        ]
+
+    @api.multi
+    def _on_right_partner(self):
         '''Determine wether functions are called in a situation where the
```
Comment (pedrobaeza) [inline partner_relations/models/res_partner_relation.py]: Double quote

File: partner_relations/models/res_partner_relation.py [hunk 8]
```diff
@@ -329,25 +293,24 @@ def _check_relation_uniqueness(self):
                 _('There is already a similar relation with overlapping dates')
             )
 
-    def get_action_related_partners(self, cr, uid, ids, context=None):
+    @api.multi
+    def get_action_related_partners(self):
         '''return a window action showing a list of partners taking part in the
```
Comment (pedrobaeza) [inline partner_relations/models/res_partner_relation.py]: Double quote

File: partner_relations/models/res_partner_relation_type_selection.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+'''
+from psycopg2.extensions import AsIs
+from openerp import api, fields, models
+from openerp.tools import drop_view_if_exists
+from .res_partner_relation_type import ResPartnerRelationType
+from .res_partner import PADDING
+
+
+class ResPartnerRelationTypeSelection(models.Model):
+    '''Virtual relation types'''
```
Comment (pedrobaeza) [inline partner_relations/models/res_partner_relation_type_selection.py]: Double quote

**Comment by pedrobaeza** [comment]
Minor nitpickings, but :+1:

**Comment by hbrunn** [comment]
@NL66278 merged your proposals

**Comment by hbrunn** [comment]
@NL66278 yay, green

**Comment by NL66278** [comment]
:+1:  I think this is really ready to be merged now.

**Comment by pedrobaeza** [comment]
Can you please squash a bit your commits and I'll merge.

**Comment by hbrunn** [comment]
@pedrobaeza done. I'd like to keep @NL66278's backported changes fom v9 untouched so that we can still cherry pick back and forth as/if necessary

**Comment by pedrobaeza** [comment]
A little last one pylint error:

```
partner_relations/models/res_partner_relation_all.py:351:17: E124 closing bracket does not match visual indentation
```

**Comment by hbrunn** [comment]
@pedrobaeza thanks, done

File: partner_relations/models/res_partner_relation_all.py [hunk 10]
```diff
@@ -0,0 +1,442 @@
...
+                if 'right_partner_id' in vals:
+                    left_partner_id = vals['right_partner_id']
+                    del vals['right_partner_id']
+                if left_partner_id:
+                    vals['left_partner_id'] = left_partner_id
+                if right_partner_id:
+                    vals['right_partner_id'] = right_partner_id
+        return vals
+
+        def check_type_selection_domain(type_selection_domain):
```
Comment (yvaucher) [inline partner_relations/models/res_partner_relation_all.py]: I guess the `return vals` should be insinde the if.

partner_relations/models/res_partner_relation_all.py:316: [W0101(unreachable), ResPartnerRelationAll._correct_vals.check_type_selection_domain] Unreachable code

Comment (hbrunn) [reply partner_relations/models/res_partner_relation_all.py]: this should be quite different. Thanks for noticing! Something must have blown up during the rebasing, will investigate

**Comment by hbrunn** [comment]
@pedrobaeza @yvaucher can we merge this? Runbot dying is unrelated

**Comment by pedrobaeza** [comment]
Please squash last three commits in one (they are bound) and I'll merge.

**Comment by hbrunn** [comment]
@pedrobaeza done

**Pull request merged**